File: 0333-Largest BST Subtree/0333-Largest BST Subtree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
 class Solution {
public:
    int largestBSTSubtree(TreeNode* root) {
        if (root == nullptr) {
            return 0;
        }
        
        int maxCount = 1;
        isBST(root, maxCount);
        return maxCount;
    }
    
private:
    struct resultType {
        bool isbst;
        int count;
        int minVal;
        int maxVal;
        resultType(bool ib, int c, int minv, int maxv): isbst(ib), count(c), minVal(minv), maxVal(maxv){}
    };

    resultType isBST(TreeNode* root, int& maxCount) {
        if (root == nullptr) {
            return resultType(true, 0, INT_MAX, INT_MIN);
        }
        
        resultType leftResult = isBST(root->left, maxCount);
        resultType rightResult = isBST(root->right, maxCount);
        if (leftResult.isbst && rightResult.isbst &&
            (root->val > leftResult.maxVal || root->left == nullptr) &&
            (root->val < rightResult.minVal || root->right == nullptr)) {
            maxCount = max(maxCount, 1 + leftResult.count + rightResult.count);
            return resultType(true, 1 + leftResult.count + rightResult.count, root->left ? leftResult.minVal : root->val, 
                    root->right ? rightResult.maxVal : root->val);
        }
        else {
            return resultType(false, 0, INT_MAX, INT_MIN); 
        }
    }
};

File: 0301-0400/0333-Largest BST Subtree/0333-Largest BST Subtree.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution:
    def largestBSTSubtree(self, root: TreeNode) -> int:
        maxCount = 0
        def dfs(root):
            if root is None:
                return True, 0, None, None
            lBST, lCount, lMinVal, lMaxVal = dfs(root.left)
            rBST, rCount, rMinVal, rMaxVal = dfs(root.right)
            if (lBST and (lMaxVal is None or root.val > lMaxVal) and
                rBST and (rMinVal is None or root.val < rMinVal)):
                currCount = lCount + 1 + rCount
                nonlocal maxCount
                maxCount = max(maxCount, currCount)
                return True, currCount, lMinVal or root.val, rMaxVal or root.val
            else:
                return False, 0, None, None
        dfs(root)
        return maxCount
